### DISCUSSION TITLE
I published small P-I-D control library in micropython for esp32 ( for other platforms should work too):

### DISCUSSION THREAD

**Discussion by Elektuur** [micropython/micropython#15975] [General]
I published small P-I-D control library in micropython for esp32 ( for other platforms should work too):
[ https://github.com/2dof/esp_control](https://github.com/2dof/esp_control)

It is based typical on functional code (as base to OOP) and provide PID algorithms and signal processing based on industrial standard implementation ( but without alarms). 
Also I added a curve generator for Setpoint generation in thermal proces control.

from my test and simulations library should be used for proces samplins Ts > 0.1 sec. 
In the furture I will add other algorithms (auto tuning mechanism, pole placement etc) and  'user-end-implementations' examples. 

I hope description is clear and someone will find it usefull.

_Originally posted by @2dof in https://github.com/micropython/micropython/discussions/10975_

**Comment by Elektuur** [comment]
Thanks for creating the Control P-I-D library. I am using it to create a PID controller. I have no experience with PID controllers and control theory, but I am studying it.
For test setup I have a K thermocouple touching an old light bulb which I use as heading element. I am using the pid_aw_updateControl() function of your library. When I enable the Integrator with pid.Ki = 0, I still get a contribution form the Integrator to the output. Why? But my problem is that the anti-windup doesn’t seem to be working. When I switch on the system it takes some time to get near the set temperature. During that time the value of pid.Ik increases and causes large overshoots. How can I prevent the overshoots?
The output of the PID is set between 0 end 230. The output is used to control a triac which controls the light bulb. The PID output value is approximately the voltage to the light bulb. The set point is for example 90 degrees C.

Image of lamp with sensor and controllor:
![20241007_123207](https://github.com/user-attachments/assets/f78d502b-ff10-427f-b036-cce70bb83dee)

**Comment by 2dof** [comment]
If You are using My code then  use:
```
pid. CFG.Isel =False 
```
to turn off I-action (and of antiwindup) - You get then only propotional action ( by default D-action is off ) 
pid.Ki is calculated in ```  pid_tune(pid)  ``` function, you can force it to overvrite but Isel field should be used to turn- on/off I-action (with antiwidup)

Antiwindup is not for prevent overshoots  it is used to prevent and reduce control value saturation and  the time in which he resides ( i.e when controller will try to put control value more than 230 value(in Your case)  then antifindup will try reduce that,  for any  control value below 230 antiwindup will not fork.
Overshooting/oscilating is caused by bad parmeters (Kp, Ti) setup  and You should use some tuning method first.
Since  pid.Ki =pid.Kp /pid.Ti  , then increasing Ti  You will decrease effect of Ki gain,  and propably overshooting,. 

for setup:  
```
pid. CFG.Isel =False 
pid. CFG.Dsel =False 
```
You get propotional controller and there will be always under/over shoot   

Now:
1. Did You performed [[Open-Loop Step Response](https://ctms.engin.umich.edu/CTMS/index.php?example=Introduction&section=ControlPID#10)](https://www.controleng.com/articles/pid-spotlight-part-7-open-loop-tuning-of-a-self-limiting-process/)   of Your process?  -> to find out  process Gain and process time ? -it is helpful in tuning , best if You can add code to safe samples ( control value, and temperature (best value from thermocouple) if not just  perform test form link .  

3. I assume that K-thermocouple measurment  unit (electronic/ linearization ) is Ok,, but at at least You should have additional temperature measutment for check and I suggest to correct your temp. measurment setup:

 - Now, Your sensor (thermocule) measure mostly air temperature near bulb (only one point of contact with bulb and rest is air near bulb)- so  real bulb temp. is higher ->  wrap some aluminum foil to get faster reaction and more accureate bulb surface temp.   


If something is not understandable - ask,

**Comment by bidrohini** [comment]
Thanks for sharing.
